spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java

```diff
@@ -0,0 +1,201 @@
+import org.springframework.util.StringUtils;
+
+import com.mongodb.WriteResult;
+
+/**
+ * MongoDbMetadataStore implementation of {@link ConcurrentMetadataStore}.
+ * Use this {@link org.springframework.integration.metadata.MetadataStore} to
+ * achieve meta-data persistence shared across application instances and
+ * restarts.
+ *
+ * @author Senthil Arumugam, Samiraj Panneer Selvam
+ * @since 4.2
+ *
+ */
+public class MongoDbMetadataStore implements ConcurrentMetadataStore {
+
+	protected final Log logger = LogFactory.getLog(getClass());
+	private final MongoTemplate template;
+	private final static String DEFAULT_COLLECTION_NAME = "metadataStore";
+	private final String collectionName;
+	private final String ID_FIELD = "_id";
+	private final String VALUE = "value";
+	
+	/**
+	 * 
+	 * Configure the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template The mongodb template
+	 * @param collectionName the collection name where it persists the data
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		Assert.notNull(template, "'template' must not be null.");
+		Assert.hasText(collectionName, "'collectionName' must not be empty.");
+		this.template = template;
+		this.collectionName = collectionName;
+	}
+
+	/**
+	 * Configure the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template The mongodb template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
+		this(template, DEFAULT_COLLECTION_NAME);
+	}
+
+	/**
+	 * Configure the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param factory The mongodb factory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory) {
+		this(factory, DEFAULT_COLLECTION_NAME);
+	}
+	
+	/**
+	 * Configure the MongoDbMetadataStore by provided {@link MongoDbFactory} and
+	 * collection name
+	 * 
+	 * @param collectionName the collection name where it persists the data
+	 * @param factory MongoDbFactory
+	 */
+	public MongoDbMetadataStore(MongoDbFactory factory, String collectionName) {
+		Assert.notNull(factory, "'factory' must not be null.");
+		Assert.hasText(collectionName, "'collectionName' must not be empty.");
+		template = new MongoTemplate(factory);
+		this.collectionName = collectionName;
+	}
+	
+	
+	/**
+	 * Method put.
+	 * @param key String
+	 * @param value String
+	 * @see org.springframework.integration.metadata.MetadataStore#put(String, String)
+	 */
+	@Override
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		Map<String, String> entry = new HashMap<String, String>();
+		entry.put(ID_FIELD, key);
+		entry.put(VALUE, value);
+		template.save(entry, collectionName);
```

Comment:
Looking at this closer I'd suggest this variant:
```
template.execute(collectionName, new CollectionCallback<Object>() {

		@Override
		public Object doInCollection(DBCollection collection) throws MongoException, DataAccessException {
			return collection.save(new BasicDBObject(entry));
		}

});
```
1. We bypass any Spring Data MongoDB stuff like events, conversions
2. We get rid of an additional `_class` document field

It isn't applied to the `putIfAbsent`, because we ill use there `script`.

BTW, please, rebase this branch to the latest master finally.